Generalize percussion set architecture

danielafeitosa opened the pull request.

danielafeitosa committed:
Load buttons from JSON file

danielafeitosa committed:
Get soundfont from config file

Resolves #6

infojunkie commented on app.js:
```diff
@@ -31,9 +31,27 @@
+const config = {}
+
+async function loadSoundFonts(instrument) {
+  let soundFont = await fetch('/soundfonts.json')
```
use `const` if the variable won't be mutated

infojunkie commented on app.js:
```diff
@@ -31,9 +31,27 @@
+async function loadInstrument(instrument) {
   return new Promise((resolve, reject) => {
-    Soundfont.instrument(new AudioContext(), 'doumbek', { soundfont: 'Doumbek-Faisal', nameToUrl: function(name, soundfont, format) {
+    let soundfont = config[instrument].soundfont
```
`const`

infojunkie commented on app.js:
```diff
@@ -44,18 +62,15 @@ async function loadInstrument() {
 }
 
 async function playButton(button) {
+  let instrument = 'doumbek';
```
Better hardcode default values as a global `const DEFAULT_INSTRUMENT`. Note you also probably need a `DEFAULT_SOUNDFONT`.

infojunkie commented on soundfonts.json:
```diff
@@ -0,0 +1,13 @@
+{
```
This file should be organized as soundfonts => instruments => mapping of MIDI keys to percussion names e.g. https://github.com/infojunkie/drumkit/blob/a95f4d242c515454e50bb778052a4246d7d8ef4d/app.js#L6-L15

The mapping of input buttons to specific MIDI keys should be kept dynamic, and defaulted with something like `const DEFAULT_BUTTONS`.

danielafeitosa committed:
Define default values as global const

Hardcode the constants:
DEFAULT_SOUNDFONT: defined as "Doumbek-Faisal"
DEFAULT_INSTRUMENT: defined as "doumbek'
DEFAULT_BUTTONS: set the buttons with the 6 first codes defined on soundfonts file

Organize the soundfonts file as suggested:
soundfonts => instruments => mapping of MIDI keys to percussion names

danielafeitosa commented:
@infojunkie updated the PR with the changes requested, please review it again :blush: 

I set the `DEFAULT_BUTTONS` with the last 6 sound codes but then I map the buttons with the first 6 codes from soundfonts. I did it to test that the codes were being updated. Let me know if I should initialize the `DEFAULT_BUTTONS` with the first 6 values

infojunkie commented on app.js:
```diff
@@ -31,9 +31,36 @@
+  return soundFonts[soundfont]
+}
+
+async function loadConfig(soundfont, instrument) {
```
I think the functions `loadSoundFonts` and `loadConfig` should be a single one that just loads the `soundfont.json` file and places it in `config`. We can then call this function when the script loads, and any subsequent function that needs soundfont information just reads from `config`.

infojunkie commented on app.js:
```diff
@@ -31,9 +31,36 @@
 async function loadInstrument() {
   return new Promise((resolve, reject) => {
-    Soundfont.instrument(new AudioContext(), 'doumbek', { soundfont: 'Doumbek-Faisal', nameToUrl: function(name, soundfont, format) {
+    Soundfont.instrument(new AudioContext(), DEFAULT_INSTRUMENT, { soundfont: DEFAULT_SOUNDFONT, nameToUrl: function(name, soundfont, format) {
```
It does not make sense to use the defaults in this function. We can define global `soundfont` and `instrument` variables that represent the current selection, and that are initialized to the defaults.

infojunkie commented on app.js:
```diff
@@ -31,9 +31,36 @@
 // 74 - Soft Tek Front
 
 let drums = null;
+let buttons = null;
```
I think this should be initialized after `DEFAULT_BUTTONS` and set as a copy of it.

infojunkie commented on app.js:
```diff
@@ -43,19 +70,26 @@ async function loadInstrument() {
   });
 }
 
+function mapButtons(availableButtons) {
```
If `buttons` is set to a copy of `DEFAULT_BUTTONS`, then we can defer the mapping from buttons to instrument sounds to some future UI, and this function is no longer needed.

infojunkie commented on app.js:
```diff
@@ -43,19 +70,26 @@ async function loadInstrument() {
-    'b4': 63,
-    'b5': 64,
-    'b6': 65,
+  let percussionSet = await loadConfig(DEFAULT_SOUNDFONT, DEFAULT_INSTRUMENT);
```
These 4 lines are unneeded if we make the changes outlined above. Essentially, this function only does the following:
- loads the drums if not loaded
- plays the sound corresponding to the already mapped button

infojunkie commented on app.js:
```diff
@@ -31,9 +31,36 @@
 // 74 - Soft Tek Front
```
We can remove the Doumbek comments since we moved the relevant info to `soundfonts.json`.

infojunkie commented:
Thanks for these changes :heart: Left some more comments :sweat_smile:

danielafeitosa committed:
Fix PR comments
